### PR TITLE
[CDAP-5955] - Improve UI build time

### PR DESCRIPTION
**Pull request by ajainarayanan**
## TL;DR -
  - Improve UI build there by the SDK build. 
  - Current solution shaves off roughly 5 mins from the sdk build (with all mvn, npm & bower caches).

### Existing problems:
  - Currently we have npm and bower that takes time to pull in all the node modules
  - The existing multi-config webpack build is really slow. We are un-necessarily building libraries (running them through babel) and building UI code for no reason.
  - (Trivial) Current incremental builds takes so much time for each file saves

### Resolutions in this PR:
- Parallelize multi-config webpack builds 
- Cache pre-built library files (as webpack dlls). This would help us avoid building it every single time as this is going to be updated less comparatively to the application code

### Analysis:
  
graph

screen shot 2017-01-05 at 3 11 03 pm

```
Maven Standalone SDK Build
-----------------------------

New Build Process:
 
  [INFO] BUILD SUCCESS
  [INFO] ------------------------------------------------------------------------
  [INFO] Total time: 21:02 min
  [INFO] Finished at: 2017-01-05T14:08:36-08:00
  [INFO] Final Memory: 265M/827M
  [INFO] ------------------------------------------------------------------------

Old Build Process:

  [INFO] ------------------------------------------------------------------------
  [INFO] BUILD SUCCESS
  [INFO] ------------------------------------------------------------------------
  [INFO] Total time: 24:32 min
  [INFO] Finished at: 2017-01-05T15:53:56-08:00
  [INFO] Final Memory: 238M/754M
  [INFO] ------------------------------------------------------------------------

```

### Note about Yarn:
Currently we are not using `Yarn` yet as `yarn --prod` doesn't clean up development dependencies properly and this adds upto 100mb to the build size. Since we ship node modules with SDK and want to reduce the sdk size, we are right now not using `yarn`.

The corresponding issue can be tracked here (https://github.com/yarnpkg/yarn/issues/761). Once this is fixed we should be able to able to replace npm with yarn and have more performance benefits.

### CODE REVIEW

**Comment by elfenheart** [comment]
Can you add a README on how to update the DLLs?

**Comment by elfenheart** [comment]
Code looks good, but I think we should wait until the SASS migration PR is merged, since the DLL have to be regenerated..

**Comment by ajainarayanan** [comment]
@elfenheart We don't need to regenerate the dlls. The dlls simply contain pre-built js library files. Are there any css styles that we want to pre-build?

**Comment by elfenheart** [comment]
I think you have to add the dlls to the POM so it gets copied to SDK or RPM build

**Comment by elfenheart** [comment]
👍  LGTM
